File: cpp/876._Middle_of_the_Linked_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        ListNode* middle = head;
        ListNode* end = head;
        while (end != NULL and end->next != NULL) {
            middle = middle->next;
            end = end->next->next;
        }
        return middle;
    }
};

File: js/876_Middle_of_the_Linked_List.js
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var middleNode = function(head) {
    let middle = head;
    let end = head;
    while (end !== null && end.next !== null) {
        middle = middle.next;
        end = end.next.next;
    }
    return middle;
};